cmake: New target which generates a sort of development kit for llext

edersondisouza opened the pull request:
Loadable extensions need access to Zephyr (and Zephyr application) includes and some CFLAGS to be properly built. This RFC adds a new target, `llext-edk`, which generates a tar file with those includes and flags that can be loaded from make and make files.

This tar file can then be used by extension developers to build extensions that can be loaded by the Zephyr application - it becomes some sort of - for a better name, suggestions welcome =D - Extension Development Kit or EDK.

The main point of the EDK is to provide an interface for extension developers, which may not have access to the Zephyr application source code, so their build is entirely out of tree. The expected usage flow would be like:

1. Application developer creates an application, with its own API in addition to that of Zephyr;
2. Application developer builds the EDK via `west build -t llext-edk`;
3. Application developer somehow makes the EDK available to extension developer;
4. Extension developer extracts EDK and includes it in its build system - the EDK provides some files to aid getting CFLAGS for cmake or make;
5. Extension developer can build the extension.

The final form of the extension or how it is loaded by the application is beyond the scope of the EDK - it's a "contract" between application and extension developers. That said, I can see how in the future some features for this could be included in Zephyr and the EDK.

Last patch in this series is an example of an application and a few extensions. The README contains the instructions for building everything, and should give a good feeling of how everything fits together. Ideally, this sample would live in several repositories: one for the application, and one for each extension, to show how the EDK acts as the bridge to build the extensions. But to ease review (and because having several repositories is too much), it's another sample.

As with any RFC, comments/critics/suggestions are welcome =D

Note 1: One of the patches is actually taken from https://github.com/zephyrproject-rtos/zephyr/pull/69540 - and it's here because it's needed for the sample to work. Once that PR is in, I'll rebase this one.
Note 2: The sample uses `EXPORT_SYMBOL` to export some Zephyr APIs to the extensions (like semaphores and threads). However, `checkpatch.pl` complains that the exports should be close to the symbols being exported. I didn't go changing several places in Zephyr just for this, to avoid distraction during review, but I can update the PR to do that. Not sure how Zephyr will handle this: will every API be exported at some point?

edersondisouza committed:
cmake: New target which generates a sort of development kit for llext

Loadable extensions need access to Zephyr (and Zephyr application)
includes and some CFLAGS to be properly built. This patch adds a new
target, `llext-edk`, which generates a tar file with those includes and
flags that can be loaded from cmake and make files.

A Zephyr application willing to expose some API to extensions it loads
only need to add the include directories describing such APIs to the
Zephyr ones via zephyr_include_directories() CMake call.

A new Kconfig option, CONFIG_LLEXT_EDK_NAME allows one to control some
aspects of the generated file, which enables some customization - think
of an application called ACME, willing to have a ACME_EXTENSION_KIT or
something.

All EDK Kconfig options are behind CONFIG_LLEXT_EDK, which doesn't
depend on LLEXT directly - so that EDK features can be leveraged by
downstream variations of loadable extensions.

Also, each arch may need different compiler flags for extensions: those
are handled by the `LLEXT_CFLAGS` cmake flag. An example is set for GCC
ARM.

Finally, EDK throughout this patch means Extension Development Kit,
which is a bad name, but at least doesn't conflict with SDK.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
subsys/llext: Generate syscalls stubs for EDK exclusively for userspace

A new Kconfig option which generates syscall stubs assuming that
extensions will always run on userspace, thus simplifying linking
them, as there's no need for z_impl_ stubs (used for direct syscalls),
CONFIG_LLEXT_EDK_USERSPACE_ONLY.

While defining __ZEPHYR_USER__ could have the same effect for optmised
builds, people building extensions on debug environments - thus
non-optimised - would suffer, as they'd need to somehow make the stubs
available (by either exporting the symbol or implementing dummy stubs).

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
cmake/llext-edk: Generate more granular CFLAGS

Besides the LLEXT_CFLAGS, which have all that is needed to compile,
generate more granular ones, LLEXT_INCLUDE_CFLAGS,
LLEXT_ALL_INCLUDE_CFLAGS, LLEXT_GENERATED_INCLUDE_CFLAGS and
LLEXT_BASE_CFLAGS. These are done for convenience, as they can help on
different setups, such as unit testing.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
tests/misc: Add llext EDK test

This test uses pytest to generate an EDK from a simple Zephyr
application, and uses this EDK to build a simple extension, to ensure
that EDK generation is sane.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
samples/subsys/llext: EDK sample

Shows a simple application which loads extensions and some simple
extensions. While everything is inside Zephyr tree, everything can
actually be build from different directories (even machines), as long as
the EDK is generated from the application and used by the extensions.
More information is available at sample's README.

This sample is build only for twister, as it requires a few steps to be
properly run, namely build the EDK, install it somewhere, build the
extensions using the EDK and finally build the application with the
extensions.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

cfriedt reviewed (commented):
Love it!

teburd commented on samples/subsys/llext/edk/app/src/main.c:
```diff
@@ -0,0 +1,188 @@
+/* checkpatch.pl complains about these EXPORT_SYMBOL - ideally, they should
+ * be close to the symbol, not sneakily exported.
+ */
+EXPORT_SYMBOL(z_arm_thread_is_in_user_mode);
```
These exports probably should exist, there’s the export_syscall macro as well to make atleast those look nicer

edersondisouza replied:
Had some issues with this one, if you could please look at this patch: https://github.com/zephyrproject-rtos/zephyr/pull/69831/commits/9db9a15b25fe2c0fe13f3e62b7bbadd649959349 (`syscalls: Introduce __syscall_generated`).
If willing to not follow the spirit of `checkpatch.pl` rule here, I guess that using `EXPORT_SYSCALL` would trick it, and allowing us export the symbol directly in the application.

teburd replied:
Actually that's pretty clever, maybe instead of having EXPORT_SYSCALL at all use the gen_syscall.py script to generate all exports for all syscalls? Is there actually a syscall we *wouldn't* want to export?

edersondisouza replied:
Hmmm - my latest headache was when some subsystem that needed syscalls were not built. But maybe just EXPORT_SYSCALL would be enough to prevent syscalls being optimised away - will do a test here.
Note that some symbols which are not syscalls would need to be exported via EXPORT_SYMBOL anyway (like `z_arm_thread_is_in_user_mode`).

edersondisouza replied:
Give a try at this idea at #70324

teburd commented on samples/subsys/llext/edk/ext1/CMakeLists.txt:
Could we reuse any of the CMake functions for llext already available by perhaps exporting with the edk a common llext.cmake file?

teburd reviewed (commented):
This looks like a really high quality pr and feature add. Adding the exports as proper exports seems like an easy improvement. Reuse of the cmake work @pillo79 has done would be great to see.

fantastic stuff!

pillo79 commented on cmake/llext-edk.cmake:
```diff
@@ -0,0 +1,106 @@
+# Generate flags for Makefile
+list(APPEND all_flags_make ${base_flags_make} ${inc_gen_flags_make} ${inc_flags_make})
+list(JOIN all_flags_make " " all_flags_str)
+file(WRITE ${LLEXT_EDK}/Makefile.cflags "LLEXT_CFLAGS = ${all_flags_str}")
```
Have you considered using the approach used by the [makefile_exports](https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/makefile_exports/CMakeLists.txt) target? I think that's clearer and also should allow greater flexibility in generating the flag lists.

edersondisouza replied:
Yep, that was my first try - it become something completely different because I wanted to trim details from the environment where the EDK is built: as the CFLAGS have absolute paths, they would leek into the EDK. That's why there's the "preprocessing" bit before, removing those. Thoughts?

pillo79 reviewed (commented):
First, a huge thank you for working on this - both the EDK infrastructure and the sample code are neatly done, documented and very welcome additions. Learned a lot myself! :smile:

My only issue with this PR is that it exports a different set of flags than what Zephyr's `add_llext_target` uses, this should be addressed IMO. 
* On this topic, would it be possible to encapsulate that logic and share it between Zephyr and the EDK code? 
(Note: before doing this I suggest to wait until https://github.com/zephyrproject-rtos/zephyr/pull/67997 gets merged, which is hopefully soon, to avoid overlapping refactorings).

The rest are just my ideas for improvement and/or curiosity for details!

edersondisouza commented on cmake/compiler/gcc/target_arm.cmake:
```diff
@@ -60,3 +60,9 @@ set(LLEXT_APPEND_FLAGS
+    -mlong-calls
+    -mthumb
+    -nodefaultlibs
+    -c)
```
As in whoever is building the extension using the EDK is responsible for adding it?

edersondisouza commented on samples/subsys/llext/edk/ext1/toolchain.cmake:
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+
+set(CMAKE_C_COMPILER   arm-zephyr-eabi-gcc)
+set(CMAKE_FIND_ROOT_PATH $ENV{ZEPHYR_SDK_INSTALL_DIR}/arm-zephyr-eabi)
```
You mean, include the compiler itself? If so, that seem dangerous - what if some application developer is using some proprietary toolchain? I agree that the toolchain needs to be the same for application and extension developers, but that's part of their "contract". 
Or are you talking about something else?

edersondisouza replied:
Hmmm - but that would assume Zephyr SDK is in use (due `ZEPHYR_SDK_INSTALL_DIR` env var). This won't always be the case, I kinda feel this would be confusing - to ship the EDK with a flag that doesn't always make sense.

edersondisouza commented on CMakeLists.txt:
```diff
@@ -2007,6 +2007,43 @@ if((CMAKE_BUILD_TYPE IN_LIST build_types) AND (NOT NO_BUILD_TYPE_WARNING))
+            ${PYTHON_EXECUTABLE}
+            ${ZEPHYR_BASE}/scripts/build/gen_syscalls.py
+            --json-file        ${syscalls_json}                     # Read this file
+            --base-output      include/generated/syscalls           # Write to this dir
+            --syscall-dispatch include/generated/syscall_dispatch.c # Write this file
```
Nice idea, will try.

edersondisouza commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,131 @@
+   :gen-args: -DSDK_BUILD=1
+   :compact:
+
+Note that the ``SDK_BUILD`` option is necessary to build the EDK initially,
+as it prevents the application from trying to load the extensions before
+they are built.
```
But if I build the EDK from another app, it won't include the application specific syscalls. Agree `SDK_BUILD` is confusing, will see if I can avoid it somehow.

edersondisouza commented:
> First, a huge thank you for working on this - both the EDK infrastructure and the sample code are neatly done, documented and very welcome additions. Learned a lot myself! 😄
> 
> My only issue with this PR is that it exports a different set of flags than what Zephyr's `add_llext_target` uses, this should be addressed IMO.
> 
> * On this topic, would it be possible to encapsulate that logic and share it between Zephyr and the EDK code?
>   (Note: before doing this I suggest to wait until [llext: additional CMake harmonizations #67997](https://github.com/zephyrproject-rtos/zephyr/pull/67997) gets merged, which is hopefully soon, to avoid overlapping refactorings).

That's certainly nice, already subscribed to #67997 =D

> 
> The rest are just my ideas for improvement and/or curiosity for details!

edersondisouza commented:
v2:
  - No more need for `SDK_BUILD` on the sample;
  - Syscalls regeneration simplified;
  - `EXPORT_SYMBOL` is now closer to symbol [1];
  - Did _not_ handle the share code with `add_llext_target` now, waiting #67997 

[1] Had some issues here when the syscall is not used by the application - it seems to be optimised away independently  from `LINKER_KEEP`. Added a patch tackling that, that if reasonable, probably deserves its own PR.

edersondisouza commented:
v3:
 - Workaround to shut up `checkpatch.pl`. Need to tackle this more properly later, it seems too much on top of this PR.

cfriedt reviewed (dismissed):
Still love it ♥️

edersondisouza commented:
I have a rebase in the works, adding DNM for now

edersondisouza committed:
cmake/modules: Extract flags filtering code to a function

This way they can be reused by the LLEXT EDK.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
cmake: Reuse llext flags for EDK

Some flags are common between in tree extensions and out of tree
supported by the EDK. Instead of duplicating those flags, the EDK reuses
the llext ones.

However, as the EDK has its own needs, two new lists,
`LLEXT_EDK_APPEND_FLAGS` and `LLEXT_EDK_REMOVE_FLAGS` are defined to
allow EDK to append or remove flags as needed.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
cmake: Handle `imacros` from flags

Instead of hardcoding `autoconf.h` imacro, get the list of imacros from
the llext flags. As those come in the form of absolute paths, they also
need to be massaged to point from the EDK directory without revealing
host complete paths.

Also, the EDK now keeps the imacros on a different flag,
`LLEXT_GENERATED_IMACROS_CFLAGS`, to keep it similar to other generated
includes.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza commented:
v3:
  - Reusing flags from internal llext support;
  - Added `llext-edk` to usage help;
  - Rebased.

Note that the reuse of llext flags used by the `add_llext_target` was done on different patches to help review. If deemed necessary, I can squash those.

andyross commented on CMakeLists.txt:
```diff
@@ -2023,6 +2024,52 @@ if((CMAKE_BUILD_TYPE IN_LIST build_types) AND (NOT NO_BUILD_TYPE_WARNING))
+            -E make_directory edk/include/generated
+        COMMAND
+            ${PYTHON_EXECUTABLE}
+            ${ZEPHYR_BASE}/scripts/build/gen_syscalls.py
```
Can't this share all the boilerplate for how to call the python script with the userspace layer?  Or better, share the targets.  It's not clear to me if an llext SDK generation ever happens in the same build as a USERSPACE=y app build, but in theory they might and if they do, we don't want to  be building everything twice.

edersondisouza replied:
I kinda expect the edk to be generated much less frequently than the app - even on CI, I would expected the edk to be generated only after the app passes all its own tests - but can't really assume that. 
That said, things are only rebuilt "needlessly" here if `CONFIG_LLEXT_EDK_USERSPACE_ONLY` is not set, as in this case we could just reuse the files built for the app. Otherwise, we need to rebuild them to skip the "inside kernel check".
I'd say that we can tackle this if this sometimes extra build becomes too burdensome - thoughts?

tejlmand replied:
Not sure I fully follow.

Is it a problem for the EDK if the syscall stubs contains routing code ?
Ref: 
> If the EDK is expected to be used by userspace only extensions, this option will make EDK stubs not contain the routing code, and only generate the userspace one.

The edk has a dependency to the Zephyr target here: 
https://github.com/zephyrproject-rtos/zephyr/blob/18090582613238a2bbee1e4c9493550453db104b/CMakeLists.txt#L2067

which means that Zephyr and thus the `gen_syscalls.py` will always be brought up-to-date before this code executes.
So exactly when can't the already generated syscall header and dispatch table just be reused ?

Does it provide a problem if the code excluded by the `gen_syscalls.py --userspace-only` flag is present ?

pillo79 commented on cmake/llext-edk.cmake:
```diff
@@ -0,0 +1,149 @@
+list(JOIN inc_gen_flags_make " " inc_gen_flags_str)
+file(APPEND ${LLEXT_EDK}/Makefile.cflags "\n\nLLEXT_GENERATED_INCLUDE_CFLAGS = ${inc_gen_flags_str}")
+
+file(APPEND ${LLEXT_EDK}/Makefile.cflags "\n\nLLEXT_BASE_CFLAGS = ${base_flags_make}")
+
+file(APPEND ${LLEXT_EDK}/Makefile.cflags "\n\nLLEXT_GENERATED_IMACROS_CFLAGS = ${imacros_gen_make}")
```
Definitely missed a list conversion for these, these do end up as a single word with semicolons in a Makefile.

pillo79 commented on cmake/llext-edk.cmake:
```diff
@@ -0,0 +1,149 @@
+list(JOIN inc_gen_flags_cmake " " inc_gen_flags_str)
+file(APPEND ${LLEXT_EDK}/cmake.cflags "\n\nset(LLEXT_GENERATED_INCLUDE_CFLAGS ${inc_gen_flags_str})")
+
+file(APPEND ${LLEXT_EDK}/cmake.cflags "\n\nset(LLEXT_BASE_CFLAGS ${base_flags_cmake})")
+
+file(APPEND ${LLEXT_EDK}/cmake.cflags "\n\nset(LLEXT_GENERATED_IMACROS_CFLAGS ${imacros_gen_cmake})")
```
Missed list conversions here as well, though in this case you could also remove them all: since you're creating CMake code semicolons should be fine :slightly_smiling_face:

pillo79 reviewed (commented):
I spotted a minor issue with list expansion, but otherwise looks great!

One general nit: both in the main `CMakeLists.txt` and in `llext-edk.cmake`, I would rename `LLEXT_CFLAGS` to `LLEXT_EDK_CFLAGS` (keeping the appropriate casing), since the same name ~is currently~ _has been_ already used in `extensions.cmake` for the actual llext build. However that confused me, not CMake... :slightly_smiling_face:

Edit: not used anymore, sleepy brain! :facepalm: But still probably worth the rename just for consistency.

teburd commented on cmake/modules/extensions.cmake:
```diff
@@ -5532,3 +5520,40 @@ function(add_llext_command)
+#   llext_filter_zephyr_flags(LLEXT_REMOVE_FLAGS zephyr_flags zephyr_filtered_flags)
+#
+function(llext_filter_zephyr_flags filter flags outvar)
+  if(NOT ${filter})
+    set(${outvar} ${flags} PARENT_SCOPE)
+    return()
+  endif()
```
Somedays I dream of using meson or... yeah anything but cmake, but then we'd have *different* problems presumably. Or maybe not.

edersondisouza commented:
v4:
  - Properly covert lists for Makefile and cmake targets;
  - Fix some variables expansion;
  - Disabled native_sim test for LLVM - it needs a bit more love, support for `--config` option and use clang properly;
  - Rebased.

pillo79 reviewed (dismissed):
LGTM, just have 2 minor non-blocking remarks.

nashif commented:
@tejlmand  can you please review

carlescufi commented:
DNM until @tejlmand has had a chance to review

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -5532,3 +5520,35 @@ function(add_llext_command)
+       REPLACE "(.+)" "^\\1$"
+       OUTPUT_VARIABLE llext_remove_flags_regexp
+  )
+  string(REPLACE ";" "|" llext_remove_flags_regexp "${llext_remove_flags_regexp}")
```
I know this is moved code, but it caught my attention now.

This form of replacement is actually risky because it also replaces an escaped `\;` with a `\|` even though an escaped `;` is not separating items in a CMake list.

pillo79 replied:
Not sure why I originally went down that road. Perhaps this alternative is better:
```suggestion
  list(JOIN llext_remove_flags_regexp "|" llext_remove_flags_regexp)
```

teburd replied:
This can be done in a follow up imho

edersondisouza replied:
Yep - indeed, as this bug is present on current code, I've created another [PR](https://github.com/zephyrproject-rtos/zephyr/pull/72437) to tackle that - as this one can still drag for some time.

tejlmand commented on cmake/compiler/gcc/target_arm.cmake:
```diff
@@ -59,8 +59,11 @@ set(LLEXT_APPEND_FLAGS
-set(LLEXT_CFLAGS
-    -mlong-calls
-    -mthumb
+list(APPEND LLEXT_EDK_REMOVE_FLAGS
+    --sysroot=.*
+    -fmacro-prefix-map=.*
+    )
+
+list(APPEND LLEXT_EDK_APPEND_FLAGS
```
`LLEXT_*` settings has been misplaced in earlier PR, but let's not continue bad structure of introducing llext_edk_*` settings.

tejlmand replied:
I understand your line of thought, but the toolchain infrastructure, and thus the `cmake/compiler/gcc/*` is for setting up the toolchain for building Zephyr. Something which is done in the current CMake invocation.

Where those flags are for exporting when building the extension, which is something quite different.

There is no obvious place for an alternative location atm.
I did consider if llext should have its own CMake module, but that collides a bit with the fact that a llext-edk target in now created.

Accepted to keep for now, but could you file an enhancement issue so that we don't forget a later cleanup, as well as providing a bit of history as to why this location was accepted. So that we can reference in future PRs.

tejlmand commented on cmake/llext-edk.cmake:
```diff
@@ -4,25 +4,7 @@
-    if (NOT EXISTS ${dir})
-        continue()
-    endif()
+function(relative_dir dir relative_out bindir_out)
```
please describe the function.

tejlmand commented on cmake/llext-edk.cmake:
```diff
@@ -53,17 +35,69 @@ foreach(dir ${INCLUDE_DIRS})
 
     cmake_path(RELATIVE_PATH dest BASE_DIRECTORY ${LLEXT_EDK} OUTPUT_VARIABLE dest_rel)
-    if(to_prj_bindir)
+    if(bindir)
```
is it correctly understood that this `bindir` is used to ensure that the include paths from the build dir comes first in the list of includes ?

If we are not in control / aligned on the include path order when creating the edk then how are we sure that the other include paths are ordered correctly / ordered the same as they were for the Zephyr build ?

And if they are ordered identical to the Zephyr build, then why do we need special treatment of the build dir ?

edersondisouza replied:
Not really - it's only to have the granular flags distinguish between flags that point to generated files and not generated ones.

tejlmand replied:
and why is that distinction important ?

There is a risk that a changed include order can impact which header files are being picked up.

For a simple sample like `hello_world` for native_sim, then Zephyr uses this order:
```
/<path-to-zephyr>/include                                                                           
/<path-to-zephyr>/build/zephyr/include/generated                                                    
/<path-to-zephyr>/soc/native/inf_clock                                                              
/<path-to-zephyr>/boards/native/native_sim                                                          
/<path-to-zephyr>/scripts/native_simulator/common/src/include                                       
/<path-to-zephyr>/scripts/native_simulator/native/src/include 
```
but with this code, that now becomes two settings, like this:
```
LLEXT_INCLUDE_CFLAGS = 
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/include                                                   
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/soc/native/inf_clock                                      
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/boards/native/native_sim                                  
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/scripts/native_simulator/common/src/include               
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/scripts/native_simulator/native/src/include
```

```
LLEXT_GENERATED_INCLUDE_CFLAGS =
$(LLEXT_EDK_INSTALL_DIR)/include/zephyr/include/generated
```

Meaning order will be changed. Most likely not an issue for this small example, but a more complex application could be affected.

I would personally prefer that include order is kept, just to not having the risk of a change in order causes hard to debug issues.

edersondisouza replied:
> and why is that distinction important ?

The granular CFLAGS help for use cases such as unit tests - you could mock the generated stuff, for instance. I expect most of the time people will only use `LLEXT_CFLAGS` with the whole set of flags.

> I would personally prefer that include order is kept, just to not having the risk of a change in order causes hard to debug issues.

Fair. I've updated the PR so that for the whole set of flags, the includes order isn't changed.

tejlmand commented on samples/subsys/llext/edk/app/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+list(APPEND LLEXT_EDK_REMOVE_FLAGS -mcpu=cortex-r5 -mfloat-abi=hard)
```
doesn't sound right to me to remove float abi flags.
What will happen later when loading the llext library ?

This seems like a recipe for a ABI incompatibility disaster.

edersondisouza replied:
The sample actually works without those, but not with them - maybe this is a bug somewhere else? qemu? toolchain? llext loader? I've added a comment about why it's needed on the file.

tejlmand replied:
> The sample actually works without those, but not with them 

How do you know ?

Have you ensured that your test sample actually uses any floating point instruction or is it just a dummy sample which happens to work because it doesn't use floatingpoint at all ?

So far i'm not concerned that we are just pushing some code which happens to work in a small testcase, but which may actually cause some hard to debug issues later.

I need a bit more explanation of why this is needed or a finer restriction for platforms that breaks.
So far this is not sufficient imo:
```
# A permission fault happens on qemu_cortex_r5 when using these
# flags. Remove them for now while we figure out the issue.
```

edersondisouza replied:
So I did a bit more investigation around this, and it seems that it's related to some relocation instructions not currently supported by llext - and removing those flags happen to cause the compiler to not generate code using those instructions, hence why the sample does work without them.  Created a new issue for that at #72832 - and noted that at the comment for the hack.

Note that this workaround is applied only for the sample - it's not done for the EDK itself - and should be revisited when issue is fixed.

tejlmand replied:
> Note that this workaround is applied only for the sample - it's not done for the EDK itself - and should be revisited when issue is fixed.

I know, but it's the only example in-tree we will have, so users will look at this and use it as reference code.
Thus a bit of extra care must be taken.

Thanks for the additional input, comment, and reference to the opened issue 👍

tejlmand commented on samples/subsys/llext/edk/app/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+list(APPEND LLEXT_EDK_REMOVE_FLAGS -mcpu=cortex-r5 -mfloat-abi=hard)
+cmake_minimum_required(VERSION 3.13.1)
```
```suggestion
cmake_minimum_required(VERSION 3.20)
```

edersondisouza commented on cmake/llext-edk.cmake:
```diff
@@ -0,0 +1,79 @@
+file(ARCHIVE_CREATE
+    OUTPUT ${LLEXT_EDK_FILE}
+    PATHS ${LLEXT_EDK}
+    FORMAT gnutar
+    COMPRESSION XZ
```
Maybe? It should be simple enough to also add a .zip generation if there's demand for that.

edersondisouza commented on scripts/build/gen_syscalls.py:
```diff
@@ -424,6 +431,8 @@ def parse_args():
                         help="Generate marshalling files (*_mrsh.c)")
     parser.add_argument("-e", "--syscall-export-llext",
                         help="output C system call export for extensions")
+    parser.add_argument("-u", "--userspace-only", action="store_true",
+                        help="Only generate the userpace path of wrappers")
```
It is - with the routing code, the `z_impl_*` will be undefined symbols on builds where only userspace matter. This can be a problem on some toolchains, so the application developer has to stub them. If the symbol is not there to start with, this issue is gone.

edersondisouza commented on cmake/llext-edk.cmake:
```diff
@@ -0,0 +1,79 @@
+list(JOIN all_flags_make " " all_flags_str)
+file(WRITE ${LLEXT_EDK}/Makefile.cflags "LLEXT_CFLAGS = ${all_flags_str}")
+
+file(WRITE ${LLEXT_EDK}/cmake.cflags "set(LLEXT_CFLAGS ${all_flags_cmake})")
```
That's neat!

edersondisouza commented:
v5:
  - Documentation for cmake/llext-edk file;
  - Documentation for cmake function;
  - Fix documentation issues;
  - Ensure cmake required version is `3.20.0`;
  - llext-edk target not behind CONFIG_LLEXT_EDK anymore;
  - CONFIG_LLEXT_EDK is gone, now all CONFIG_LLEXT_EDK_* are behind a `menu`;
  - Fix issue when not using `west`;
  - Use `CMAKE_CURRENT_LIST_DIR` for cmake.cflags;
  - Uniform naming of variables (use lower case for local ones);
  - Fix typo on commit message;
  - Rebased.

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,126 @@
+.. zephyr:code-sample:: llext-edk
+   :name: Linkable loadable extensions EDK
+
```
```suggestion
.. zephyr:code-sample:: llext-edk
   :name: Linkable loadable extensions EDK
   :relevant-api: llext

   Enable linkable loadable extension development outside the Zephyr tree using LLEXT EDK (Extension Development Kit)
```

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,126 @@
+providing some level of isolation - although the kernel one still has access
+to all of Zephyr kernel.
+
+Note that the kernel extension is only available when the EDK is built with
+``CONFIG_LLEXT_EDK_USERSPACE_ONLY=n``.
```
```suggestion
Note that the kernel extension is only available when the EDK is built with the
:kconfig:`CONFIG_LLEXT_EDK_USERSPACE_ONLY` option disabled.
```

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,126 @@
+
+This sample demonstrates how to use the Zephyr LLEXT EDK (Extension Development
+Kit). It is composed of one Zephyr application, which provides APIs for the
+extensions that it loads. The API provided is a simple publish/subscribe system,
```
```suggestion
extensions that it loads. The provided API is a simple publish/subscribe system,
```

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,126 @@
+This sample demonstrates how to use the Zephyr LLEXT EDK (Extension Development
+Kit). It is composed of one Zephyr application, which provides APIs for the
+extensions that it loads. The API provided is a simple publish/subscribe system,
+based on Zbus, which extensions use to communicate with each other.
```
```suggestion
based on :ref:`Zbus <zbus>`, which extensions use to communicate with each other.
```

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,126 @@
+Building the extensions
+***********************
+
+First, ensure the Zephyr SDK location is available at the environment variable
+``ZEPHYR_SDK_INSTALL_DIR``:
+
+.. code-block:: console
+
+    export ZEPHYR_SDK_INSTALL_DIR=</path/to/zephyr-sdk>
+
+As the ``ZEPHYR_SDK_INSTALL_DIR`` is used by the extensions to find the Zephyr SDK.
+
```
```suggestion
The :envvar:`ZEPHYR_SDK_INSTALL_DIR` environment variable is used by the
extensions to find the Zephyr SDK, so you need to ensure it's properly set:

.. code-block:: console

    export ZEPHYR_SDK_INSTALL_DIR=</path/to/zephyr-sdk>

```

kartben reviewed (changes requested):
Requesting a few changes in the README of the sample, if you don't mind

I would really like to see the EDK mentioned in the [main LLEXT doc page](https://docs.zephyrproject.org/latest/services/llext/index.html) too at some point, but this can come later :) In fact, some of the text that's been used to document this sample is effectively the documentation for the EDK itself, so maybe some of it can be moved to the main doc page at some point.

Thanks for the great work!

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,132 @@
+   :relevant-api: llext
+
+
+Enable linkable loadable extension development outside the Zephyr tree using
+LLEXT EDK (Extension Development Kit).
```
Needs 3-space indent as this belongs to the code-sample directive

kartben commented on samples/subsys/llext/edk/README.rst:
```diff
@@ -0,0 +1,132 @@
+.. code-block:: console
+
+    export ZEPHYR_SDK_INSTALL_DIR=</path/to/zephyr-sdk>
+
+As the ``ZEPHYR_SDK_INSTALL_DIR`` is used by the extensions to find the Zephyr SDK.
```
```suggestion
```

edersondisouza commented:
v6:
  - Sample README updates.

kartben reviewed (dismissed):
+1 for docs / sample, thanks!

teburd commented:
@tejlmand I marked some comments from you, myself, and others that were clearly resolved. The remaining unresolved comments seem to be primarily questions or comments on cmake that was there prior to this PR. Please un-resolve if you feel the issues have not yet been resolved.

nashif commented:
@tejlmand  please revisit

edersondisouza commented:
v7:
  - Rebased.

tejlmand reviewed (commented):
my main concern still lies with the comment here: https://github.com/zephyrproject-rtos/zephyr/pull/69831#discussion_r1601230814

I need some more info / investigation on this.
How much has this been tested / verified ?
Especially, has the loadable extension been tested with floating point instructions ?

I need a bit more detailed info on exactly what / why this happens before approving something which just removes ABI flags.

edersondisouza commented:
v8:
  - Preserve order of includes on generated flags;
  - Allow one to use `-DLLEXT_EDK_INSTALL_DIR` directly on CMake for extensions (in addition to env var);
  - Better describe the workaround needed to run the application.

tejlmand reviewed (approved):
Approved.

Still concerned on the removal, but at least the code is better commented and there is a followup issue, #72832.
Currently the best that can be done, while also allowing the feature to be merged and get more user feedback.
